master/buildbot/reporters/telegram.py

```diff
@@ -0,0 +1,1055 @@
+
+    def get(self, item, default=None):
+        return self.chat.get(item, default)
+
+    def update(self, src):
+        self.chat.update(src)
+
+    def list_notified_events(self):
+        if self.notify_events:
+            self.send("The following events are being notified:\n{}"
+                      .format("\n".join(sorted(
+                          "🔔 **{}**".format(n) for n in self.notify_events))))
+        else:
+            self.send("🔕 No events are being notified.")
+
+
+def collect_fields(fields):
+    for field in fields:
+        if field['fullName']:
+            yield field
+        if 'fields' in field:
+            yield from collect_fields(field['fields'])
+
+
+class TelegramContact(Contact):
+
+    def __init__(self, bot, user=None, channel=None):
+        super().__init__(bot, user, channel)
+        self.template = None
+
+    @property
+    def chatid(self):
+        return self.channel.id
+
+    @property
+    def userid(self):
+        if isinstance(self.user, dict):
+            return self.user['id']
+        else:
+            return self.user
+
+    @property
+    def user_full_name(self):
+        fullname = " ".join((self.user['first_name'],
+                             self.user.get('last_name', ''))).strip()
+        return fullname
+
+    @property
+    def user_name(self):
+        return self.user['first_name']
+
+    def describeUser(self):
+        if not isinstance(self.user, dict):
+            return self.user
+
+        user = self.user_full_name
+        try:
+            user += ' (@{})'.format(self.user['username'])
+        except KeyError:
+            pass
+
+        if self.channel.id != self.userid:
+            chat_title = self.channel.get('title')
+            if chat_title:
+                user += " on '{}'".format(chat_title)
+
+        return user
+
+    _scared_users = {}
+
+    _stop_phrase = (
+        ("👾  Do it not, you can!  👾", 'CAADAgADAxgAAkKvaQABG4A6r70tTawWBA'),
+        ("👹  You shall not pass!!!  👹", 'CAADAgAD1AIAAmMr4gkRoBV--rBVehYE'),
+    )
+
+    @defer.inlineCallbacks
+    def access_denied(self, *args, tmessage, **kwargs):
+        uid = self.user['id']
+        if uid == self.channel['id']:
+            text, sticker = random.choice(self._stop_phrase)
+            now = util.now()
+            # clean users scared some time ago
+            horizon = now - 120
+            for u, t in list(self._scared_users.items()):
+                if t < horizon:
+                    del self._scared_users[u]
+            if self._scared_users.get(uid) is None:
+                self._scared_users[uid] = now
+                yield self.send(
+                    "{}".format(text),
+                    reply_to_message_id=tmessage['message_id'])
+                yield self.bot.send_sticker(uid, sticker)
+                return
+
+        self.send(
+            "⛔  ACCESS DENIED  ⛔\n\n" +
+            random.choice((
+                "You are simply not allowed to to this!",
```

Comment:
I think it's worth refraining from adding new jokes to the codebase. In this case we are adding a sizeable amount of code for no real benefit for the end user.

Reply:
I think you are right. Actually the IRC bot has way too much of such stuff (I have already moved most of it from `words.py` to `irc.py`.

I have removed the complicated code, however, I have left several _access denied_ messages, so there is more variety in the chat. I prefer that the bot mimics human conversation rather than be just a command line. This is why I tried to make it answer in complete sentences where applicable.